packages/core/src/transports/offline.ts

```diff
@@ -0,0 +1,169 @@
+}
+
+interface OfflineStore {
+  insert(env: Envelope): Promise<void>;
+  pop(): Promise<Envelope | undefined>;
+}
+
+export type CreateOfflineStore = (options: OfflineTransportOptions) => OfflineStore;
+
+export interface OfflineTransportOptions extends InternalBaseTransportOptions {
+  /**
+   * A function that creates the offline store instance.
+   */
+  createStore?: CreateOfflineStore;
+
+  /**
+   * Flush the offline store shortly after startup.
+   *
+   * Defaults: false
+   */
+  flushAtStartup?: boolean;
+
+  /**
+   * Called before an event is stored.
+   *
+   * Return false to drop the envelope rather than store it.
+   *
+   * @param envelope The envelope that failed to send.
+   * @param error The error that occurred.
+   * @param retryDelay The current retry delay in milliseconds.
+   */
+  shouldStore?: (envelope: Envelope, error: Error, retryDelay: number) => boolean | Promise<boolean>;
+}
+
+type Timer = number | { unref?: () => void };
+
+/**
+ * Wraps a transport and stores and retries events when they fail to send.
+ *
+ * @param createTransport The transport to wrap.
+ */
+export function makeOfflineTransport<TO>(
+  createTransport: (options: TO) => Transport,
+): (options: TO & OfflineTransportOptions) => Transport {
+  return options => {
+    const transport = createTransport(options);
+    const store = options.createStore ? options.createStore(options) : undefined;
+
+    let retryDelay = START_DELAY;
+    let flushTimer: Timer | undefined;
+
+    function shouldQueue(env: Envelope, error: Error, retryDelay: number): boolean | Promise<boolean> {
+      // We don't queue Session Replay envelopes because they are:
+      // - Ordered and Replay relies on the response status to know when they're successfully sent.
+      // - Likely to fill the queue quickly and block other events from being sent.
+      if (isReplayEnvelope(env)) {
+        return false;
+      }
+
+      if (options.shouldStore) {
+        return options.shouldStore(env, error, retryDelay);
+      }
+
+      return true;
+    }
+
+    function flushIn(delay: number): void {
+      if (!store) {
+        return;
+      }
+
+      if (flushTimer) {
+        clearTimeout(flushTimer as ReturnType<typeof setTimeout>);
+      }
+
+      flushTimer = setTimeout(async () => {
+        flushTimer = undefined;
+
+        const found = await store.pop();
+        if (found) {
+          log('Attempting to send previously queued event');
+          void send(found).catch(e => {
+            log('Failed to retry sending', e);
+          });
+        }
+      }, delay) as Timer;
+
+      // We need to unref the timer in node.js, otherwise the node process never exit.
+      if (typeof flushTimer !== 'number' && typeof flushTimer.unref === 'function') {
+        flushTimer.unref();
+      }
+    }
+
+    function flushWithBackOff(): void {
+      if (flushTimer) {
+        return;
+      }
+
+      flushIn(retryDelay);
+
+      retryDelay *= 2;
```

Comment:
l: Could be rewritten as:

```js
retryDelay = Math.min(retryDelay * 2, MAX_DELAY);
```

Maybe a tick shorter?

Reply:
Yes!